Code/GraphMol/SubstructLibrary/SubstructLibrary.h

```diff
@@ -0,0 +1,310 @@
+/*!
+    This implementation uses quite a bit less memory than the
+    cached binary or uncached implementation.  However, due to the
+    reduced speed it should be used in conjunction with a pattern
+    fingerprinter.
+
+    See RDKit::FPHolder
+*/
+class CachedSmilesMolHolder: public MolHolderBase {
+  std::vector<std::string> mols;
+public:
+  CachedSmilesMolHolder() : MolHolderBase(), mols() {}
+
+  virtual unsigned int addMol( const ROMol &m ) {
+
+    bool doIsomericSmiles = true;
+    mols.push_back(MolToSmiles(m, doIsomericSmiles));
+    return size()-1;
+  }
+  
+  virtual boost::shared_ptr<ROMol> getMol(unsigned int idx) const {
+    // throws IndexError?
+    boost::shared_ptr<ROMol> mol(SmilesToMol(mols[idx]));
+    return mol;
+  }
+  
+  virtual unsigned int size() const { return rdcast<unsigned int>(mols.size()); }
+
+};
+
+
+//! Base FPI for the fingerprinter used to rule out impossible matches
+class FPHolderBase {
+  std::vector<ExplicitBitVect *> fps;
+
+public:
+ virtual ~FPHolderBase() {
+   for(size_t i=0;i<fps.size();++i)
+     delete fps[i];
+   
+ }
+
+ //! Add a molecule to the fingerprinter
+ unsigned int addMol( const ROMol &m) {
+   fps.push_back(getQueryBits(m));
+   return rdcast<unsigned int>(fps.size() - 1);
+ }
+
+ //! Return false if a substructure search can never match the molecule
+ bool passesFilter(unsigned int idx, const ExplicitBitVect &query) const {
+   PRECONDITION(idx < fps.size(), "idx out of range in PatternHolder::passesFilter");
+   return AllProbeBitsMatch(query, *fps[idx]);
+ }
+ 
+ virtual ExplicitBitVect *getQueryBits(const ROMol &m) const = 0;
+ 
+};
+
+//! Uses the pattern fingerprinter to rule out matches
+class PatternHolder : public FPHolderBase {
+public:
+  virtual ExplicitBitVect *getQueryBits(const ROMol &m) const {
+    return PatternFingerprintMol(m, 2048);
+  }
+};
+
+//! Substtructure Search a library of molecules
+/*!  This class allows for multithreaded substructure searches os
+     large datasets.
+
+     The implementations can use fingerprints to speed up searches
+     and have molecules cached as binary forms to reduce memory
+     usage.
+*/
+class SubstructLibrary {
+  boost::shared_ptr<MolHolderBase> molholder;
+  boost::shared_ptr<FPHolderBase> fpholder;
+  MolHolderBase *mols; // used for a small optimization
+  FPHolderBase *fps;
+public:
+  SubstructLibrary() : molholder( new MolHolder ), fpholder() {}
+  
+  SubstructLibrary( boost::shared_ptr<MolHolderBase> molecules ) :
+       molholder(molecules), fpholder(), mols(molholder.get()), fps(0) {
+  }
+
+  SubstructLibrary( boost::shared_ptr<MolHolderBase>  molecules,
+                    boost::shared_ptr<FPHolderBase> fingerprints ) :
+       molholder(molecules), fpholder(fingerprints),
+       mols(molholder.get()), fps(fpholder.get()) {
+  }
+      
+
+  //!Get the underlying molecule holder implementation
+        MolHolderBase & getMolHolder() { return *mols; }
```

Comment:
I'll do that, but this can never be null due to the way the class is constructed.